src/lib/style-properties.html

```diff
@@ -235,55 +241,74 @@
-      // collect any custom properties into `props`.
-      //
-      // TODO(sorvell): this should change to collecting properties from any
-      // :host(...) and then matching these against self.
-      hostPropertiesFromStyles: function(styles) {
-        if (!styles._hostStyleProperties) {
-          styles._hostStyleProperties =
-            this.selectedPropertiesFromStyles(styles, this.HOST_SELECTORS);
-        }
-        return styles._hostStyleProperties;
-      },
-
-      selectedPropertiesFromStyles: function(styles, selectors) {
-        var props = {}, self = this;
-        styleUtil.forRulesInStyles(styles, function(rule) {
+      hostAndRootPropertiesForScope: function(scope) {
+        var hostProps = {}, rootProps = {}, self = this;
+        styleUtil.forActiveRulesInStyles(scope._styles, function(rule, style) {
           if (!rule.propertyInfo) {
             self.decorateRule(rule);
           }
-          for (var i=0; i < selectors.length; i++) {
-            if (rule.parsedSelector === selectors[i]) {
-              self.collectProperties(rule, props);
-              return;
+          var hostScope = scope.is ?
+            styleTransformer._calcHostScope(scope.is, scope.extends) :
+             'html';
+          var parsedSelector = rule.parsedSelector;
+          var isRoot = parsedSelector === ':root';
+          var isHost = parsedSelector.indexOf(':host') === 0;
+          // rule._buildType is set for rules from custom-styles, or the dom-module of the element has it
+          var cssBuildType = scope.__cssBuild || style.__cssBuild;
+          if (cssBuildType === 'shady') {
+            // :root -> x-foo > *.x-foo for elements and html for custom-style
+            isRoot = parsedSelector === (hostScope + '> *.' + hostScope) || parsedSelector.indexOf('html') !== -1;
+            // :host -> x-foo for elements, but sub-rules have .x-foo in them
+            isHost = !isRoot && parsedSelector.indexOf(hostScope) === 0;
+          }
+          if (cssBuildType === 'shadow') {
+            isRoot = parsedSelector === ':host > *' || parsedSelector === 'html';
+            isHost = isHost && !isRoot;
+          }
+          if (!isRoot && !isHost) {
+            return;
+          }
+          var selectorToMatch = hostScope;
+          if (isHost) {
+            // need to transform :host under ShadowDOM because `:host` does not work with `matches`
+            if (settings.useNativeShadow && !rule.transformedSelector) {
+              // transform :host into a matchable selector
+              rule.transformedSelector =
+                styleTransformer._transformRuleCss(
+                  rule,
+                  styleTransformer._transformComplexSelector,
+                  scope.is,
+                  hostScope
+                );
+            }
+            selectorToMatch = rule.transformedSelector || hostScope;
+          }
+          if (scope && rule.propertyInfo.properties && selectorToMatch) {
```

Comment:
Why check `scope` here? Perhaps we can bail earlier if `!rule.propertyInfo.properties` (avoiding the unnecessary setup work if we don't use it)